Use std::span more in WTFString

cdumez opened the pull request:
#### 492d572f84caae7af7b09bc7b2d0da09cc6c4c1e
<pre>
Use std::span more in WTFString
<a href="https://bugs.webkit.org/show_bug.cgi?id=272067">https://bugs.webkit.org/show_bug.cgi?id=272067</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::computeCurrencySortKey):
(JSC::extractCurrencySortKey):
(JSC::computeCurrencyDigits):
* Source/WTF/wtf/text/ASCIILiteral.h:
(WTF::StringLiterals::operator _charspan):
* Source/WTF/wtf/text/AtomString.h:
(WTF::IntegerToStringConversionTrait&lt;AtomString&gt;::flush):
* Source/WTF/wtf/text/IntegerToStringConversion.h:
(WTF::numberToStringImpl):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::appendCharacters):
(WTF::IntegerToStringConversionTrait&lt;StringBuilder&gt;::flush):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::containsOnly):
(WTF::isSpecialCharacter const):
* Source/WTF/wtf/text/StringView.h:
(WTF::isSpecialCharacter const):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::ascii const):
(WTF::String::latin1 const):
* Source/WTF/wtf/text/WTFString.h:
(WTF::IntegerToStringConversionTrait&lt;String&gt;::flush):

Canonical link: <a href="https://commits.webkit.org/277043@main">https://commits.webkit.org/277043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dec3dbd30bd2a705c4a40b2f51758d8309a299f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46479 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25635 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49083 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49155 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42520 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23098 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37926 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47057 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/22661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19176 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41188 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4524 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/39724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41564 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50993 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/45958 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21484 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/17926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45170 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44108 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/53104 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6492 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22479 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10887 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
```diff
@@ -135,35 +135,35 @@ static unsigned computeCurrencyDigits(const String& currency)
-        { "XAF", 0 },
-        { "XOF", 0 },
-        { "XPF", 0 }
+    static constexpr std::pair<std::span<const char>, unsigned> currencyMinorUnits[] = {
```
Can we use a `span<char, 3>` or `std::array<char, 3>`? Won’t work with the new `_charspan`, though.

darinadler commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -381,35 +381,29 @@ CString String::ascii() const
+    auto characters = span16();
     char* characterBuffer;
-    CString result = CString::newUninitialized(length, characterBuffer);
+    CString result = CString::newUninitialized(characters.size(), characterBuffer);
```
Does this chop a `size_t` to an `unsigned`?

cdumez replied:
I guess I should update some functions to take in a `size_t` instead of an `unsigned` and add an assertion that the value fits in an unsigned if they ever store the value in an `unsigned` variable. Let me know if you had something else in mind.

cdumez replied:
newUninitialized::newUninitialized takes in a size_t already so no issue for this particular instance.

darinadler replied:
Yes, your solution sounds correct for functions that don’t already take a `size_t`. We do `RELEASE_ASSERT` in string-related functions for this. There are lots of them already. In fact there are also checks for too-large `unsigned` values, but they don’t work if a `size_t` is already chopped.

darinadler commented on Source/WTF/wtf/text/StringBuilder.h:
```diff
@@ -51,9 +51,12 @@ class StringBuilder {
+    WTF_EXPORT_PRIVATE void appendCharacters(const UChar*, unsigned); // FIXME: Port callers to pass a span and remove.
+    WTF_EXPORT_PRIVATE void appendCharacters(const LChar*, unsigned); // FIXME: Port callers to pass a span and remove.
+    void appendCharacters(const char* characters, unsigned length) { appendCharacters(reinterpret_cast<const LChar*>(characters), length); } // FIXME: Port callers to pass a span and remove.
+    void appendCharacters(std::span<const UChar> characters) { appendCharacters(characters.data(), characters.size()); }
```
This chops a `size_t` to an `unsigned`.

darinadler commented on Source/WTF/wtf/text/StringBuilder.h:
```diff
@@ -51,9 +51,12 @@ class StringBuilder {
+    WTF_EXPORT_PRIVATE void appendCharacters(const LChar*, unsigned); // FIXME: Port callers to pass a span and remove.
+    void appendCharacters(const char* characters, unsigned length) { appendCharacters(reinterpret_cast<const LChar*>(characters), length); } // FIXME: Port callers to pass a span and remove.
+    void appendCharacters(std::span<const UChar> characters) { appendCharacters(characters.data(), characters.size()); }
+    void appendCharacters(std::span<const LChar> characters) { appendCharacters(characters.data(), characters.size()); }
```
This chops a `size_t` to an `unsigned`.

darinadler commented on Source/WTF/wtf/text/StringBuilder.h:
```diff
@@ -51,9 +51,12 @@ class StringBuilder {
+    void appendCharacters(const char* characters, unsigned length) { appendCharacters(reinterpret_cast<const LChar*>(characters), length); } // FIXME: Port callers to pass a span and remove.
+    void appendCharacters(std::span<const UChar> characters) { appendCharacters(characters.data(), characters.size()); }
+    void appendCharacters(std::span<const LChar> characters) { appendCharacters(characters.data(), characters.size()); }
+    void appendCharacters(std::span<const char> characters) { appendCharacters(reinterpret_cast<const LChar*>(characters.data()), characters.size()); }
```
This chops a `size_t` to an `unsigned`.

darinadler commented on Source/WTF/wtf/text/AtomString.h:
```diff
@@ -30,8 +30,10 @@ class AtomString final {
-    AtomString(const UChar*, unsigned length);
+    AtomString(const LChar*, unsigned length); // FIXME: Port call sites to pass a span and remove.
+    AtomString(const UChar*, unsigned length); // FIXME: Port call sites to pass a span and remove.
+    AtomString(std::span<const LChar> characters) : AtomString(characters.data(), characters.size()) { }
```
This chops a `size_t` to an `unsigned`.

darinadler commented on Source/WTF/wtf/text/AtomString.h:
```diff
@@ -30,8 +30,10 @@ class AtomString final {
+    AtomString(const LChar*, unsigned length); // FIXME: Port call sites to pass a span and remove.
+    AtomString(const UChar*, unsigned length); // FIXME: Port call sites to pass a span and remove.
+    AtomString(std::span<const LChar> characters) : AtomString(characters.data(), characters.size()) { }
+    AtomString(std::span<const UChar> characters) : AtomString(characters.data(), characters.size()) { }
```
This chops a `size_t` to an `unsigned`.

darinadler commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -381,35 +381,29 @@ CString String::ascii() const
 
         char* characterBuffer;
-        CString result = CString::newUninitialized(length, characterBuffer);
+        CString result = CString::newUninitialized(characters.size(), characterBuffer);
```
Does this chop a `size_t` to an `unsigned`?

darinadler replied:
It does not.

darinadler commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -419,23 +413,18 @@ CString String::latin1() const
+    auto characters = this->span16();
     char* characterBuffer;
-    CString result = CString::newUninitialized(length, characterBuffer);
+    CString result = CString::newUninitialized(characters.size(), characterBuffer);
```
Does this chop a `size_t` to an `unsigned`?

darinadler replied:
It does not.

darinadler reviewed (changes requested):
Need to make sure we check lengths before chopping them to a smaller integer size.

darinadler commented on Source/WTF/wtf/text/StringBuilder.cpp:
```diff
@@ -149,28 +149,30 @@ UChar* StringBuilder::extendBufferForAppendingWithUpconvert(unsigned requiredLen
     }
-    if (auto destination = extendBufferForAppendingWithUpconvert(saturatedSum<uint32_t>(m_length, length)))
-        StringImpl::copyCharacters(destination, characters, length);
+    ASSERT(characters.size() < std::numeric_limits<uint32_t>::max());
```
We’ve been using `RELEASE_ASSERT` or `CRASH` for these kinds of length checks in `String` code. We need to ask ourselves exactly what we want to happen when the total size is too big for a `String` or too big for a `uint32_t` and make sure that the code does that.

darinadler commented on Source/WTF/wtf/text/StringBuilder.cpp:
```diff
@@ -149,28 +149,30 @@ UChar* StringBuilder::extendBufferForAppendingWithUpconvert(unsigned requiredLen
-    if (auto destination = extendBufferForAppendingWithUpconvert(saturatedSum<uint32_t>(m_length, length)))
-        StringImpl::copyCharacters(destination, characters, length);
+    ASSERT(characters.size() < std::numeric_limits<uint32_t>::max());
+    if (auto destination = extendBufferForAppendingWithUpconvert(saturatedSum<uint32_t>(m_length, static_cast<uint32_t>(characters.size()))))
```
I’m not sure `saturatedSum` is acceptable here given our goals. We need to think about what happens when the sum saturates.

darinadler replied:
I figured it out, if it saturates then the buffer size will be greater than `MaxStringLength`, so the extend will fail. All is well.

darinadler commented on Source/WTF/wtf/text/StringBuilder.cpp:
```diff
@@ -149,28 +149,30 @@ UChar* StringBuilder::extendBufferForAppendingWithUpconvert(unsigned requiredLen
-    if (!length || hasOverflowed())
+    if (characters.empty() || hasOverflowed())
         return;
+    ASSERT(characters.size() < std::numeric_limits<uint32_t>::max());
```
Same comments as above apply here.

cdumez committed:
Use std::span more in WTFString
https://bugs.webkit.org/show_bug.cgi?id=272067

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::computeCurrencySortKey):
(JSC::extractCurrencySortKey):
(JSC::computeCurrencyDigits):
* Source/WTF/wtf/text/ASCIILiteral.h:
(WTF::StringLiterals::operator _charspan):
* Source/WTF/wtf/text/AtomString.h:
(WTF::IntegerToStringConversionTrait<AtomString>::flush):
* Source/WTF/wtf/text/IntegerToStringConversion.h:
(WTF::numberToStringImpl):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::appendCharacters):
(WTF::IntegerToStringConversionTrait<StringBuilder>::flush):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::containsOnly):
(WTF::isSpecialCharacter const):
* Source/WTF/wtf/text/StringView.h:
(WTF::isSpecialCharacter const):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::ascii const):
(WTF::String::latin1 const):
* Source/WTF/wtf/text/WTFString.h:
(WTF::IntegerToStringConversionTrait<String>::flush):

Canonical link: https://commits.webkit.org/277043@main
